core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala

```diff
@@ -35,48 +31,36 @@ class SnapshotService(protected val userInfo: UserInfo, val dataSource: SlickDat
-            } yield { bigQueryRef }
-
-            createBQReferenceFuture.recover {
-              case t: Throwable =>
-                //fire and forget these undos, we've made our best effort to fix things at this point
-                for {
-                  // TODO: AS-724 is it worth deleting the companion and its ref here? If the user attempted to add a snapref,
-                  // they are likely to do so again, and we can just reuse whatever companion was already created.
-                  // This will prevent continual creation/deletion of the same-named dataset.
-                  _ <- deltaLayer.deleteDataset(workspaceContext)
-                  _ <- Future(workspaceManagerDAO.deleteDataRepoSnapshotReference(workspaceContext.workspaceIdAsUUID, referenceId, userInfo.accessToken))
-                } yield {}
-                throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError, s"Unable to create snapshot reference in workspace ${workspaceContext.workspaceId}. Error: ${t.getMessage}"))
-            }
-          } else {
-            Future(workspaceManagerDAO.getBigQueryDatasetReferenceByName(workspaceContext.workspaceIdAsUUID, datasetId.getDataset, userInfo.accessToken))
-            // retrieve existing dataset reference
+          // something went wrong creating the companion dataset
+          logger.warn(s"Error creating Delta Layer companion dataset for workspace $workspaceName: ${t.getMessage}")
+          // since companion dataset creation failed, try to clean up this snapshot reference so the user can try again
+          Try(workspaceManagerDAO.deleteDataRepoSnapshotReference(wsid, snapshotRef.getMetadata.getResourceId, userInfo.accessToken)) match {
+            case Success(_) =>
+              throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError,
+                s"Unable to create snapshot reference in workspace ${workspaceContext.workspaceId} due to problems creating " +
+                  s"the Delta Layer companion dataset. Error: [${t.getMessage}]"))
+            case Failure(ex) =>
+              throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError,
+                s"Error while creating snapshot reference in workspace ${workspaceContext.workspaceId}. Additionally, there " +
+                  s"was an error cleaning up the snapshot reference. The reference may be in an unusable state. Original error " +
+                  s"during Delta Layer companion dataset creation was: [${t.getMessage}]. " +
+                  s"Error during snapshot reference cleanup: [${ex.getMessage}]."))
```

Comment:
previous code turned the cleanup into a Future and called it asynchronously; I turned it back into a synchronous operation so we can offer more precise error messages to the user.